src/BlockTypeRegistry.cpp

```diff
@@ -0,0 +1,229 @@
+
+#include "Globals.h"
+#include "BlockTypeRegistry.h"
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// BlockInfo:
+
+BlockInfo::BlockInfo(
+	const AString & aPluginName,
+	const AString & aBlockTypeName,
+	std::shared_ptr<cBlockHandler> aHandler,
+	const std::map<AString, AString> & aHints,
+	const std::map<AString, BlockInfo::HintCallback> & aHintCallbacks
+):
+	mPluginName(aPluginName),
+	mBlockTypeName(aBlockTypeName),
+	mHandler(aHandler),
+	mHints(aHints),
+	mHintCallbacks(aHintCallbacks)
+{
+}
+
+
+
+
+
+AString BlockInfo::hintValue(
+	const AString & aHintName,
+	const BlockState & aBlockState
+)
+{
+	// Search the hint callbacks first:
+	auto itrC = mHintCallbacks.find(aHintName);
+	if (itrC != mHintCallbacks.end())
+	{
+		// Hint callback found, use it:
+		return itrC->second(mBlockTypeName, aBlockState);
+	}
+
+	// Search the static hints:
+	auto itr = mHints.find(aHintName);
+	if (itr != mHints.end())
+	{
+		// Hint found, use it:
+		return itr->second;
+	}
+
+	// Nothing found, return empty string:
+	return AString();
+}
+
+
+
+
+
+void BlockInfo::setHint(const AString & aHintKey, const AString & aHintValue)
+{
```

Comment:
What if the hint was already set from a callback? Currently the value would be silently ignored when `hintValue` would be called as the callback variant is checked first.

Reply:
The callbacks are considered higher priority, so they "overwrite" anything set statically.

Yeah, I know, it's not 100% flexible, but it's good enough for our purposes and easy enough to both implement and understand.

Reply:
I've added mentions of this behavior in the doxy-comments.

Reply:
Should we also print a message when `setHint` is called for something that is already done using a callback? Otherwise someone might be confused why his hint isn't being used. Or is that overkill?

Reply:
That would require us to call the callback to know whether it provides the value or not. Definitely overkill.

Reply:
Took a second look, a warning is a cheap way to improve later coding, so I added that.